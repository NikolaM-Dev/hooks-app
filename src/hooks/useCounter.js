import { useState } from 'react';

const useCounter = ({ initialState = 10 } = {}) => {
  const [counter, setCounter] = useState(initialState);

  const increment = ({ factor = 1 }) => setCounter((state) => state + factor);

  const decrement = ({ factor = 1 }) => setCounter((state) => state - factor);

  const reset = () => setCounter(initialState);

  return {
    counter,
    increment,
    decrement,
    reset,
  };
};

export default useCounter;
