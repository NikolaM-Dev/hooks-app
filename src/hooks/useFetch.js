import { useEffect, useState } from 'react';

const useFetch = ({ url }) => {
  const [state, setState] = useState({
    data: null,
    error: null,
    loading: true,
  });

  useEffect(() => {
    setState({
      data: null,
      error: null,
      loading: true,
    });
    fetch(url)
      .then((res) => res.json())
      .then((data) =>
        setState({
          data,
          error: null,
          loading: false,
        })
      )
      .catch({
        data: null,
        error: true,
        loading: false,
      });
  }, [url]);

  return state;
};

export default useFetch;
