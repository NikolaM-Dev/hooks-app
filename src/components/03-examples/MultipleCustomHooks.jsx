import useCounter from 'hooks/useCounter';
import useFetch from 'hooks/useFetch';
// import useFetch from 'hooks/useFetch';

const MultipleCustomHooks = () => {
  const { counter, increment, decrement } = useCounter({ initialState: 1 });
  const { data, loading } = useFetch({
    url: `https://www.breakingbadapi.com/api/quotes/${counter}`,
  });

  const { quote, author } =
    !loading && data.length > 0
      ? data[0]
      : { quote: 'Quote not found', author: 'NN' };

  return (
    <div>
      <h1>BreakingBad Quotes</h1>
      <hr />
      {loading && (
        <div className="alet alert-info text-center py-3">Loading...</div>
      )}

      {!loading && (
        <section>
          <blockquote className="blockquote text-end">
            <p className="mb-3">{quote}</p>
            <footer className="blockquote-footer">{author}</footer>
          </blockquote>
          <div className="d-flex justify-content-between">
            <button onClick={decrement} className="btn btn-primary">
              Previus quote
            </button>
            <button onClick={increment} className="btn btn-primary">
              Next quote
            </button>
          </div>
        </section>
      )}
    </div>
  );
};

export default MultipleCustomHooks;
