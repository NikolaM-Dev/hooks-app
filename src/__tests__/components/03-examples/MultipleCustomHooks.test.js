import '@testing-library/jest-dom';
import { shallow } from 'enzyme';

import MultipleCustomHooks from 'components/03-examples/MultipleCustomHooks';
import useFetch from 'hooks/useFetch';
import useCounter from 'hooks/useCounter';

jest.mock('hooks/useFetch');
jest.mock('hooks/useCounter');

describe('Tests on <MultipleCustomHooks /> component', () => {
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 1,
      increment: () => {},
    });
  });

  it('should show <MultipleCustomHooks /> correctly', () => {
    useFetch.mockReturnValue({
      data: null,
      error: null,
      loading: true,
    });
    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should show the information', () => {
    const dataInfo = { author: 'Nikola', quote: 'Hello World!' };
    useFetch.mockReturnValue({
      data: [dataInfo],
      error: null,
      loading: false,
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    const isLoading = wrapper.find('.alert').exists();
    const author = wrapper.find('.blockquote-footer').text().trim();
    const quote = wrapper.find('.mb-3').text().trim();

    expect(isLoading).toBeFalsy();
    expect(author).toBe(dataInfo.author);
    expect(quote).toBe(dataInfo.quote);
  });
});
