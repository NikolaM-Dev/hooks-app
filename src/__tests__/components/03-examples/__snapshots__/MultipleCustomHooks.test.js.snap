// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <MultipleCustomHooks /> component should show <MultipleCustomHooks /> correctly 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alet alert-info text-center py-3"
  >
    Loading...
  </div>
</div>
`;
